### PR TITLE
FLINK-35641 ParquetSchemaConverter supports required fields

### PR DESCRIPTION
**Pull request by Gerrrr**
## What is the purpose of the change

The purpose of this change is to fix 2 scenarios where Flink produces incorrect Parquet files.

Scenario 1: Convert nullable Flink types to optional Parquet types. Right now, Flink configures all Parquet types as optional, regardless of whether they are nullable or not.
Scenario 2:.Ensures that the converter does not create invalid Parquet files with optional map keys. According to [Parquet standard](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#maps), map keys are required.

## Brief change log

- Configure non-nullable Flink types as required Parquet types.
- Ensure that Flink map key type is non-nullable / required.
- Ensure that Flink multiset element type is non-nullable / required.
- Adjust existing tests and add new ones to cover the change in behavior.
- Mention the nullable key limitation in Parquet format docs.

## Verifying this change

1. Adjusted existing tests.
2. Added new test cases to cover invalid types.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: yes? This change affects the behavior of the file system connector when if it writes Parquet files.

## Documentation

  - Does this pull request introduce a new feature?  no
  - If yes, how is the feature documented? Mentioned the unsupported nullable map keys in docs

### CODE REVIEW

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java [hunk 1]
```diff
@@ -143,6 +146,13 @@ private static Type convertToParquetType(
                         convertToParquetType(LIST_ELEMENT_NAME, arrayType.getElementType(), conf));
             case MAP:
                 MapType mapType = (MapType) type;
+                Preconditions.checkState(
+                        !mapType.getKeyType().isNullable(),
```
Comment (JingsongLi) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: I have concern here, does it will break the compatibility before?

**Review by JingsongLi — COMMENTED** [review]
Thanks @Gerrrr for the contribution, left one comment.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java (hunk 1 above)
Comment (Gerrrr) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: Hi @JingsongLi !

Yes, it does break backwards compatibility. After upgrade, jobs running with nullable map keys and multiset items in their schema will throw `InvalidStateException`. I do not see an alternative though as Parquet files these jobs produce are invalid and may not be readable by other tools.

**Comment by Gerrrr** [comment]
Test failures are caused by this patch - https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=60364&view=logs&j=2e8cb2f7-b2d3-5c62-9c05-cd756d33a819&t=2dd510a3-5041-5201-6dc3-54d310f68906. I will adjust the test cases on Thursday.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java (hunk 1 above)
Comment (JingsongLi) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: Can we automatically convert to not null? I understand that Flink SQL defaults to nullable.

Comment (Gerrrr) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: My reasoning about this issue is the following. 

Flink type system is fundamentally incompatible with Parquet type system. Flink allows nullable map keys, Parquet does not allow them.  Right now, Flink's `ParquetSchemaConverter` happily maps any Flink row type to Parquet message type, producing invalid Parquet files for incompatible schemas. Some external tools either silently return incorrect result or fail on these invalid Parquet files.

I see 2 possible solutions:
1. Do not allow to create tables with schema that is not compatible with Parquet spec (i.e., what's implemented in this patch).
2. Unconditionally specify that Parquet map key is `required` and throw in the writer if a record does have `null` key.

The main advantage of Solution 1 is that we correctly use Flink's type system guardrail against unsupported data types. This is what the type systems are for!
The main disadvantage of Solution 1 is that it is inconvenient to upgrade to because one'd have to update table schemas. While inconvenient, it does make the operator think whether that table can have invalid records.

The main advantage of Solution 2 is that only jobs producing invalid data will throw an exception, albeit at processing time. 
The disadvantages of this approach are:
1.  incompatible jobs will crashloop at processing time rather than at schema modeling time.
2. there will be additional per-record overhead for schemas with maps to ensure that they do not contain `null` keys. I can imagine this overhead being non-trivial, especially for complex schemas with multiple nesting levels.

For completeness, I did not bring up a solution where we unconditionally specify Parquet map key as `required`, but do not verify records in the writer, because it does not solve the problem at hand. With that appeoach, Flink still can produce invalid Parquet files, and external tooling will crash while reading them.

---

To sum up, I am fine with either of 2 solutions with the slight preference to Solution 1 because it does not introduce performance overhead and makes operators aware of the potential type incompatibility sooner than later.

What do you think?

Comment (JingsongLi) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: I tend to lean towards the second solution, because it would be a bit frustrating if the default type (Most users are too lazy to use not null) cannot be used.

Perhaps Parquet will report an error when encountering a null value? Is there no performance loss?

Comment (JingsongLi) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: It seems that when Spark SQL creates a Map type by default, the key is not null? Key cannot have null, this is its definition.

Comment (Gerrrr) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: Yes, Spark SQL Map does not allow null keys, that's why they do not have this issue. 

> Perhaps Parquet will report an error when encountering a null value?

Yes, Parquet writer will enter a crashloop on a record with null key.

Comment (Gerrrr) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: >  Is there no performance loss?

we'll have to identify maps in the schema and then check for each record that it does not have a null key. This may add up for complex nested schemas.

Comment (Gerrrr) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: I am happy to implement Solution 2 though.

Comment (JingsongLi) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java:150]: Sounds good to me.

Comment (Gerrrr) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java]: Done in https://github.com/apache/flink/pull/24956/commits/80b0bec76ee06bf8e48227d3a9f6c71ac0d3a8e6.

**Review by JingsongLi — APPROVED** [review]
Thanks @Gerrrr for update! Looks good to me!
